### PR TITLE
Add Pre-/Post-Benchmark Hooks for Plug-ins

### PR DESCRIPTION
**Pull request by dey4ss**
First, this PR adds optional hooks to the plug-in interface that are called before and after benchmark executions. For instance, these hooks can be used for applying physical designs (tiering, indexes, compression, data dependencies) before the items are executed or for exporting workload information after benchmark runs.

The `BenchmarkRunner`'s benchmark item runner is passed as an argument to the pre-benchmark hooks. Thus, plug-ins can execute the items, run advisors, etc. The post-benchmark hook gets the JSON report as argument to add or modify metrics.

Second, this PR gets rid of `constant_mappings` and the extern bimaps to map enum types to strings. We use `magic_enum::enum_name()` wherever possible. If string representations do not match the enum names, this PR moves the bimaps to the corresponding header (but *not* as `extern`). The reason is that we link other executables againt the test target and `extern` definitions of said bimaps caused segfaults in the CI coverage build (presumably multiple deallocations).

### CODE REVIEW

File: src/benchmarklib/benchmark_runner.cpp [hunk 1]
```diff
@@ -7,19 +7,18 @@
 #include <boost/accumulators/accumulators.hpp>
 #include <boost/accumulators/statistics/mean.hpp>
 #include <boost/accumulators/statistics/stats.hpp>
-
 #include <boost/algorithm/string/join.hpp>
 #include <boost/range/adaptors.hpp>
-#include "cxxopts.hpp"
+#include <cxxopts.hpp>
```
Comment (Bouncner) [inline src/benchmarklib/benchmark_runner.cpp:12]: Actually, I think boost's usage of `<>` is rather wrong. For `cxxopts` it is most certainly wrong as it is not a system library.

We might be better of pulling boost as a third party library (different boost version on Ubuntu/MacOS brew can be a pain). Then the entire discussio would be rather simple.

File: src/benchmarklib/encoding_config.cpp [hunk 2]
```diff
@@ -1,6 +1,6 @@
 #include "encoding_config.hpp"
 
-#include "constant_mappings.hpp"
+#include <magic_enum.hpp>
```
Comment (Bouncner) [inline src/benchmarklib/encoding_config.cpp:3]: See above.

File: src/lib/storage/vector_compression/compressed_vector_type.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#include "compressed_vector_type.hpp"
+
+namespace hyrise {
+
+std::ostream& operator<<(std::ostream& stream, const CompressedVectorType compressed_vector_type) {
+  switch (compressed_vector_type) {
+    case CompressedVectorType::FixedWidthInteger4Byte: {
```
Comment (Bouncner) [inline src/lib/storage/vector_compression/compressed_vector_type.cpp:7]: Can't we use `magic_enum` here as well?

File: src/lib/storage/vector_compression/compressed_vector_type.hpp [hunk 4]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <cstdint>
+#include <optional>
 
 #include <boost/version.hpp>
 #if BOOST_VERSION < 107100                 // TODO(anyone): remove this block once Ubuntu ships boost 1.71
```
Comment (Bouncner) [inline src/lib/storage/vector_compression/compressed_vector_type.hpp:7]: Just checked: Ubuntu 20 ships boost 1.71 (https://launchpad.net/ubuntu/focal/+package/libboost-dev). Do we want to support older Ubuntu versions? I'd say no.

File: src/lib/utils/meta_tables/meta_log_table.cpp [hunk 5]
```diff
@@ -1,6 +1,7 @@
 #include "meta_log_table.hpp"
 
-#include "constant_mappings.hpp"
+#include <magic_enum.hpp>
```
Comment (Bouncner) [inline src/lib/utils/meta_tables/meta_log_table.cpp:3]: See above.

File: src/lib/utils/meta_tables/segment_meta_data.cpp [hunk 6]
```diff
@@ -1,5 +1,7 @@
 #include "segment_meta_data.hpp"
 
+#include <magic_enum.hpp>
```
Comment (Bouncner) [inline src/lib/utils/meta_tables/segment_meta_data.cpp:3]: See above.

File: src/lib/utils/plugin_manager.cpp [hunk 7]
```diff
@@ -74,16 +74,25 @@ void PluginManager::load_plugin(const std::filesystem::path& path) {
...
+  const auto& post_benchmark_hook = plugin_ref.post_benchmark_hook();
+  if (post_benchmark_hook) {
+    _post_benchmark_hooks[plugin_name] = *post_benchmark_hook;
+  }
 }
 
 void PluginManager::exec_user_function(const PluginName& plugin_name, const PluginFunctionName& function_name) {
   Assert(_user_executable_functions.contains({plugin_name, function_name}),
-         "There is no " + function_name + " defined for plugin " + plugin_name + ".");
+         "There is no " + function_name + " defined for plugin `" + plugin_name + ".");
```
Comment (Bouncner) [inline src/lib/utils/plugin_manager.cpp:95]: Either remove or change to straight quotes with an end quote as well.

File: src/lib/utils/plugin_manager.cpp [hunk 8]
```diff
@@ -95,6 +104,39 @@ void PluginManager::exec_user_function(const PluginName& plugin_name, const Plug
       LogLevel::Info);
 }
 
+void PluginManager::exec_pre_benchmark_hook(const PluginName& plugin_name,
+                                            AbstractBenchmarkItemRunner& benchmark_item_runner) {
+  Assert(_pre_benchmark_hooks.contains({plugin_name}),
+         "There is no pre-benchmark hook defined for plugin `" + plugin_name + "`.");
```
Comment (Bouncner) [inline src/lib/utils/plugin_manager.cpp:110]: Wrong quotes. Multiple times.

File: src/lib/utils/plugin_manager.cpp [hunk 9]
```diff
@@ -95,6 +104,39 @@ void PluginManager::exec_user_function(const PluginName& plugin_name, const Plug
...
+  auto& log_manager = Hyrise::get().log_manager;
+  log_manager.add_message("PluginManager", "Called pre-benchmark hook provided by plugin `" + plugin_name + "`.",
+                          LogLevel::Info);
+}
+
+void PluginManager::exec_post_benchmark_hook(const PluginName& plugin_name) {
+  Assert(_post_benchmark_hooks.contains({plugin_name}),
+         "There is no post-benchmark hook defined for plugin `" + plugin_name + "`.");
+
+  const auto post_benchmark_hook = _post_benchmark_hooks[plugin_name];
```
Comment (Bouncner) [inline src/lib/utils/plugin_manager.cpp:124]: ref?

File: src/lib/utils/plugin_manager.cpp [hunk 10]
```diff
@@ -95,6 +104,39 @@ void PluginManager::exec_user_function(const PluginName& plugin_name, const Plug
       LogLevel::Info);
 }
 
+void PluginManager::exec_pre_benchmark_hook(const PluginName& plugin_name,
+                                            AbstractBenchmarkItemRunner& benchmark_item_runner) {
+  Assert(_pre_benchmark_hooks.contains({plugin_name}),
+         "There is no pre-benchmark hook defined for plugin `" + plugin_name + "`.");
+
+  const auto pre_benchmark_hook = _pre_benchmark_hooks[plugin_name];
```
Comment (Bouncner) [inline src/lib/utils/plugin_manager.cpp:112]: ref?

File: src/test/CMakeLists.txt [hunk 11]
```diff
@@ -268,6 +268,7 @@ set(
     # Added plugin targets so that we can test member methods without going through dlsym
     hyriseMvccDeletePlugin
     hyriseUccDiscoveryPlugin
+    # hyriseBenchmarkLib
```
Comment (Bouncner) [inline src/test/CMakeLists.txt:271]: ?

File: src/test/lib/operators/sort_test.cpp [hunk 12]
```diff
@@ -116,7 +116,7 @@ inline std::string sort_test_formatter(const testing::TestParamInfo<SortTestPara
   }
   stream << (param.input_is_reference ? "Reference" : "Data") << "Input";
   for (const auto& sort_column : param.sort_columns) {
-    stream << "Col" << sort_column.column << sort_mode_to_string.left.at(sort_column.sort_mode);
```
Comment (Bouncner) [inline src/test/lib/operators/sort_test.cpp:119]: Not yours, but that looks ugly. How does not look when printed? `coll_orderkeyAscending`?

File: src/test/lib/utils/plugin_manager_test.cpp [hunk 13]
```diff
@@ -1,5 +1,6 @@
 #include "base_test.hpp"
 
+// #include "../benchmarklib/tpch/tpch_benchmark_item_runner.hpp"
```
Comment (Bouncner) [inline src/test/lib/utils/plugin_manager_test.cpp:3]: ?

**Review by Bouncner — COMMENTED** [review]
Nothing serious.

Nice clean ups!

File: src/benchmarklib/benchmark_runner.cpp (hunk 1 above)
Comment (dey4ss) [reply src/benchmarklib/benchmark_runner.cpp:12]: Just grepped a few variations, seems we usually import magic_enum and boost with `<>` and only in a few cases with `""`. Totally fine to change this (altogether with the Boost <1.71 workaround). However, I'd move this to a new issue/PR if that's okay for you.

File: src/lib/storage/vector_compression/compressed_vector_type.cpp (hunk 3 above)
Comment (dey4ss) [reply src/lib/storage/vector_compression/compressed_vector_type.cpp:7]: Sure we can, just did not notice it.

File: src/lib/storage/vector_compression/compressed_vector_type.hpp (hunk 4 above)
Comment (dey4ss) [reply src/lib/storage/vector_compression/compressed_vector_type.hpp:7]: see https://github.com/hyrise/hyrise/pull/2522#discussion_r1091640241

File: src/lib/utils/plugin_manager.cpp (hunk 7 above)
Comment (dey4ss) [reply src/lib/utils/plugin_manager.cpp:95]: Argh, copied it from the here (and obviously not in the correct way.
https://github.com/hyrise/hyrise/blob/54770d02e097b783bde1c1924f45cca43502a4f3/src/lib/utils/plugin_manager.cpp#L94
I'd go for `'` if you agree - just like emphasizing the user-provided names in the messages.

File: src/test/CMakeLists.txt (hunk 11 above)
Comment (dey4ss) [reply src/test/CMakeLists.txt:271]: The outcommented cmake-stuff and the outcommented includes are an artifact from playing around to fix the CI stage, gonna remove.

Comment (dey4ss) [reply src/test/CMakeLists.txt:271]: Just uncommented everything again (required for testing the benchmark hooks). Let's see if it also passes CI. 🤞

File: src/benchmarklib/benchmark_runner.cpp (hunk 1 above)
Comment (dey4ss) [reply src/benchmarklib/benchmark_runner.cpp]: Taken from the [Boost wiki](https://github.com/boostorg/wiki/wiki/Getting-Started%3A-Overview#installing-boost):
> The initial cloning will take at least 45 minutes on a 3.0 mbps Internet connection and will consume roughly 1.5 GB of disk space. Cloning is only done once, so this lengthy initial time is not a long-term concern.

Are we sure we want this to be a third party library (CI builds, etc.)?

**Comment by dey4ss** [comment]
@Bouncner @mweisgut Still open to your opinion on passing `AbstractBenchmarkItemuRunner` to the `pre_benchmark_hook` (to, e.g., execute the workload in advance for workload-driven optimization plugins) and passing nothing to the `post_benchmark_hook`. Guess globally reachable stuff (like exporting the log) can be done with that.

File: src/benchmarklib/benchmark_runner.cpp (hunk 1 above)
Comment (Bouncner) [reply src/benchmarklib/benchmark_runner.cpp]: Well, that might be a good reason to keep it as a sys library. :D

File: src/test/lib/operators/sort_test.cpp (hunk 12 above)
Comment (Bouncner) [reply src/test/lib/operators/sort_test.cpp:119]: Hmm, no. But then again: it's just autogenerated test names. Feel free to resolve.
Sorry.

**Comment by Bouncner** [comment]
> @Bouncner @mweisgut Still open to your opinion on passing `AbstractBenchmarkItemuRunner` to the `pre_benchmark_hook` (to, e.g., execute the workload in advance for workload-driven optimization plugins) and passing nothing to the `post_benchmark_hook`. Guess globally reachable stuff (like exporting the log) can be done with that.

I think I need to discuss that in person. Not sure I can follow.

**Comment by dey4ss** [comment]
> I think I need to discuss that in person. Not sure I can follow.

Let's discuss in the fortnightly

**Comment by dey4ss** [comment]
As discussed in the fortnightly, pre-benchmark hooks now have access to the benchmark report JSON. Thus, I had to split the `BenchmarkRunner::write_report_to_file()` into report creation and actually writing to always run the benchmark hooks (they might not touch the JSON at all). However, we do not actually create the JSON if we do not write it and it is the plugin's reponsibility to cope with that or to crash.

File: src/benchmarklib/benchmark_runner.cpp (hunk 1 above)
Comment (dey4ss) [reply src/benchmarklib/benchmark_runner.cpp]: At least the <1.71 workaround is now fixed 🤷‍♀️

**Review by Bouncner — APPROVED** [review]
Just a potential miss of a const.

File: src/benchmarklib/benchmark_runner.cpp [hunk 14]
```diff
@@ -151,26 +152,29 @@ void BenchmarkRunner::run() {
     }
   }
 
+  // Create report.
+  const auto write_report = _config.output_file_path && (!_config.verify && !_config.enable_visualization);
+  auto report = write_report ? _create_report() : nlohmann::json{};
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/benchmarklib/benchmark_runner.cpp:157]: Cannot be `const` when the plugin shall be allowed to modify / add metrics to the report, right? We *could* return a ptr to the report which then can be `const`.

File: src/plugins/CMakeLists.txt [hunk 15]
```diff
@@ -31,11 +31,11 @@ function(add_plugin)
...
 endfunction(add_plugin)
 
-add_plugin(NAME hyriseMvccDeletePlugin SRCS mvcc_delete_plugin.cpp mvcc_delete_plugin.hpp DEPS sqlparser magic_enum gtest)
-add_plugin(NAME hyriseSecondTestPlugin SRCS second_test_plugin.cpp second_test_plugin.hpp DEPS sqlparser)
-add_plugin(NAME hyriseTestPlugin SRCS test_plugin.cpp test_plugin.hpp DEPS sqlparser)
-add_plugin(NAME hyriseTestNonInstantiablePlugin SRCS non_instantiable_plugin.cpp)
-add_plugin(NAME hyriseUccDiscoveryPlugin SRCS ucc_discovery_plugin.cpp ucc_discovery_plugin.hpp DEPS sqlparser magic_enum compact_vector)
+add_plugin(NAME hyriseMvccDeletePlugin SRCS mvcc_delete_plugin.cpp mvcc_delete_plugin.hpp DEPS sqlparser magic_enum gtest hyriseBenchmarkLib)
+add_plugin(NAME hyriseSecondTestPlugin SRCS second_test_plugin.cpp second_test_plugin.hpp DEPS sqlparser hyriseBenchmarkLib)
+add_plugin(NAME hyriseTestPlugin SRCS test_plugin.cpp test_plugin.hpp DEPS sqlparser hyriseBenchmarkLib hyriseBenchmarkLib)
```
Comment (mweisgut) [inline src/plugins/CMakeLists.txt:36]: Redundant `hyriseBenchmarkLib`. Will be fixed in #2448.

**Pull request merged**